### DISCUSSION TITLE
bughunting weirdness and other stuff

### DISCUSSION THREAD

**Discussion by revelator** [jmeubank/tdm-gcc#60] [Ideas]
New version is quite ingenious but there were a few things that needed checking.
mutex locks opening the same mutex in two places with CreateMutex is a recipee for disaster because the second lock would get ERROR_ALLREADY_EXIST and wont be given permission to lock the cleanup function. So use OpenMutex to check if the lock is still active 
and if that is the case use that else recreate it with CreateMutex. 

Sadly there are more examples of things that break with the mutex locking mechanism, for one glib2 started using mutex locks as well and the version they use deadlocks the TDM mutex causing more or less the same exception to be thrown as the example above.

Took me quite a while but here is a version of the shmem mechanism that is pretty reliable ->
[0160-shared-memory.zip](https://github.com/jmeubank/tdm-gcc/files/9734423/0160-shared-memory.zip)

it still borks at the glib2 mutex at times so i made an extra flag to be used in CFLAGS / CXXFLAGS -> _USE_SHMEM to turn the shared pointer code on or off in your project, downside is that anything compiled without setting -D_USE_SHMEM=1 will not be able to throw exceptions across dll boundaries but normal exceptions still work. This is a workaround untill we can figure out a way to protect the mutex functions from being tampered with by external sources.

**Comment by jmeubank** [comment]
CreateMutex on a named mutex that already exists, returns a handle to the existing mutex. Can you provide a sample where you proved that this didn't happen? (I.e. proves Microsoft's documentation about mutex behavior to be wrong.) Otherwise you're simply wasting effort with this.

Named mutexes cannot deadlock other mutexes, only themselves (i.e. the mutex with that unique name).

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Ralph Engels ***@***.***>
Sent: Friday, October 7, 2022 6:50:57 AM
To: jmeubank/tdm-gcc ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [jmeubank/tdm-gcc] bughunting weirdness and other stuff (Discussion #60)


New version is quite ingenious but there were a few things that needed checking.
mutex locks opening the same mutex in two places with CreateMutex is a recipee for disaster because the second lock would get ERROR_ALLREADY_EXIST and wont be given permission to lock the cleanup function. So use OpenMutex to check if the lock is still active
and if that is the case use that else recreate it with CreateMutex.

Sadly there are more examples of things that break with the mutex locking mechanism, for one glib2 started using mutex locks as well and the version they use deadlocks the TDM mutex causing more or less the same exception to be thrown as the example above.

Took me quite a while but here is a version of the shmem mechanism that is pretty reliable ->
0160-shared-memory.zip<https://github.com/jmeubank/tdm-gcc/files/9734423/0160-shared-memory.zip>

it still borks at the glib2 mutex at times so i made an extra flag to be used in CFLAGS / CXXFLAGS -> _USE_SHMEM to turn the shared pointer code on or off in your project, downside is that anything compiled without setting -D_USE_SHMEM=1 will not be able to throw exceptions across dll boundaries but normal exceptions still work. This is a workaround untill we can figure out a way to get protect the mutex functions from being tampered with by external sources.

**Reply by revelator** [reply]
would have to make a dump of the older version but it does return ERROR_ALLREADY_EXIST and that is a fact unfortunatly :(, i also thought that a named mutex would be safe but turns out that is not the case. An easier way is to build one of the later versions of glib2 > 3.34 then build gdk-pixbuf2 with drmingw installed as the jit debugger, you will notice it crashes when loading the image plugins with a handle to gmutex_lock / gmutex_unlock and if you load it up in gdb you will see that it deadlocks the mutex in gcc.

**Reply by revelator** [reply]
btw i just noticed that i misplaced the decrement counter it should actually only decrement if we could release the mutex as it is it only decrements if we fail to release it XD doh... been at it for days so i was a bit tired when i wrote that.

**Reply by jmeubank** [reply]
> would have to make a dump of the older version but it does return ERROR_ALLREADY_EXIST and that is a fact unfortunatly

It sets that value in GetLastError(), but still returns a handle to the mutex as desired. Note that the original code doesn't care about GetLastError(), only about whether a handle is returned.

> i also thought that a named mutex would be safe but turns out that is not the case. An easier way is to build one of the later versions of glib2 > 3.34 then build gdk-pixbuf2 with drmingw installed as the jit debugger, you will notice it crashes when loading the image plugins with a handle to gmutex_lock / gmutex_unlock and if you load it up in gdb you will see that it deadlocks the mutex in gcc.

Did you check the call stack and can you provide it?

There could very well be a synchronicity problem in the original code, but this seems like something else. A lot of your code seems like you were trying to debug an issue you ran into (which is great!) but then you left it in place even though it's not really necessary (which makes it hard for me to read and understand).

**Reply by revelator** [reply]
old version is long gone im afraid :S so that would be hard unless i rebuild it from scratch.
yeah also my thought that it should be ignored but it caused WaitForSingleObject to return ACCESS_DENIED whenever i hit the ERROR_ALLREADY_EXIST and that unfortunatly broke things.
some of the code i left in there just tries to lock down any attempt at modifying the mutex -> eg. increment counter to 1 after we got the mutex then prevent anything from trying to do that again untill we successfully released it again by decrementing the counter by 1.

also it seems you were trying to make the mutex in Shmem_NameSpace( region_struct ) *Shmem_NameSpace( init_local_region )( void ) unique with this block of code ->

```
  /* Initialize the mutex name as a printable representation of GetCurrentProcessId() + Shmem_Version_Prefix */
  const DWORD pid = GetCurrentProcessId();

  /* modifying a value that should have been a const sigh...
     this is also one of the reasons we can not get away with locking the mutex twice,
     since anything we do here ends up being reflected in the cleanup mutex.
     glib2's mutex handler breaks on this code, for some stupid reason it signales this mutex... */
  for( size_t i = 0; i < sizeof( DWORD ); ++i )
    {
      mutex_name[i * 2] = ( *( ( const unsigned char * )&pid + i ) >> 4 ) + 0x41;
      mutex_name[i * 2 + 1] = ( *( ( const unsigned char * )&pid + i ) & 0xF ) + 0x61;
    }

  memcpy0( mutex_name + ( sizeof( DWORD ) * 2 ), Shmem_Version_Prefix, sizeof( Shmem_Version_Prefix ) );
```

i guess your idea was to have two named mutexes with a slight difference so that they would not interfere with eachother but as it turns out the mutex name leaks into the one in -> int Shmem_NameSpace( cleanup_local_region )( void ). So my guess was either wrong or it was unforeseen.

**Reply by revelator** [reply]
Ill try to make a build of the old version to debug against :) would be nice to put this bug in the grave.

**Reply by revelator** [reply]
well seems i have good news, after fixing my little whoopsie with the counter lock it now builds a working gdk-pixbuf2 :O ill have to test it with other glib based packages but it seems the guard helped.

**Reply by revelator** [reply]
in case you were wondering i got the hint for the mutex access mechanism from here https://stackoverflow.com/questions/54152660/createmutex-seems-to-acquire-mutex-even-when-error-already-exists specifically this part of the discussion ->

`If the mutex is a named mutex and the object existed before this function call, the return value is a handle to the existing object, GetLastError returns ERROR_ALREADY_EXISTS, bInitialOwner is ignored, and the calling thread is not granted ownership. However, if the caller has limited access rights, the function will fail with ERROR_ACCESS_DENIED and the caller should use the OpenMutex function.`

this was actually pointed out by microsoft themself though you have to read the link in the thread a couple of times to get the idea (on first glance i was also thinking WHAAAT xD)

**Reply by jmeubank** [reply]
For what it's worth, the point of the mutex is to protect against what is a very rare corner case, where multiple DLLs containing the libgcc code are loaded in different threads at the same time. The mutex protects the linked list, composed of all shmem local regions in all DLLs with the same version and process space, from the race condition where multiple threads that load (or unload) the DLL try to modify the list simultaneously. The same mutex name is needed both at load and at unload in order to protect the same shared linked list. The vast majority of programs will not load DLLs in different threads, so the vast majority of programs do not need this mutex.

Since you seem to be writing some sort of defensive code (not good) to try and avoid the intended behavior and change it to something else, it's likely you've created something that simply stops the mutex from working as intended, and therefore bypasses whatever bug is causing problems with gdk-pixbuf2.

I am still interested in seeing a stack trace of the issue, if you can reproduce it. Also, I don't see in the Microsoft documentation under what scenario ERROR_ACCESS_DENIED is returned, either, so would like to understand that more as well. Other than that, I'm not interested in this patchset -- it seems to simply rewrite to CamelCase and a bunch of defensive programming that obfuscates instead of providing value.

**Reply by revelator** [reply]
aye i would also like it to be simpler, as is this is just a first attempt to get around some specific problems that turned up.
bootstrapping a build of the unmodified version and ill post the stack trace as soon as i can.

**Reply by revelator** [reply]
should not be hard to reproduce it as it happens 100% of the time so lets see :)

**Reply by revelator** [reply]
```
CREATE_PROCESS PID=25104 TID=27140 lpBaseOfImage=0000000000400000 \\?\I:\Msys2\tdm32\bin\gdk-pixbuf-query-loaders.exe
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00007FFA28970000 \\?\C:\Windows\System32\ntdll.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000077440000 \\?\C:\Windows\SysWOW64\ntdll.dll
CREATE_THREAD PID=25104 TID=3744
CREATE_THREAD PID=25104 TID=23088
CREATE_THREAD PID=25104 TID=17764
CREATE_THREAD PID=25104 TID=20260
CREATE_THREAD PID=25104 TID=30424
CREATE_THREAD PID=25104 TID=22420
CREATE_THREAD PID=25104 TID=19412
CREATE_THREAD PID=25104 TID=26772
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00007FFA269D0000 \\?\C:\Windows\System32\wow64.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00007FFA27320000 \\?\C:\Windows\System32\wow64win.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000077430000 \\?\C:\Windows\System32\wow64cpu.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075410000 \\?\C:\Windows\SysWOW64\kernel32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076400000 \\?\C:\Windows\SysWOW64\KernelBase.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000751F0000 \\?\C:\Windows\SysWOW64\apphelp.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000758D0000 \\?\C:\Windows\SysWOW64\msvcrt.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000064BC0000 \\?\I:\Msys2\tdm32\bin\libglib-2.0-0.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075340000 \\?\C:\Windows\SysWOW64\advapi32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000761C0000 \\?\C:\Windows\SysWOW64\sechost.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000077240000 \\?\C:\Windows\SysWOW64\rpcrt4.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076C40000 \\?\C:\Windows\SysWOW64\ole32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075720000 \\?\C:\Windows\SysWOW64\ucrtbase.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076990000 \\?\C:\Windows\SysWOW64\combase.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076C10000 \\?\C:\Windows\SysWOW64\gdi32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076870000 \\?\C:\Windows\SysWOW64\win32u.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076310000 \\?\C:\Windows\SysWOW64\gdi32full.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075610000 \\?\C:\Windows\SysWOW64\msvcp_win.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076680000 \\?\C:\Windows\SysWOW64\user32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075990000 \\?\C:\Windows\SysWOW64\shell32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000762A0000 \\?\C:\Windows\SysWOW64\ws2_32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000061EC0000 \\?\I:\Msys2\tdm32\bin\libintl-8.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000064580000 \\?\I:\Msys2\tdm32\bin\libpcre2-8-0.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000066200000 \\?\I:\Msys2\tdm32\bin\libiconv-2.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000615C0000 \\?\I:\Msys2\tdm32\bin\libgio-2.0-0.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000066E40000 \\?\I:\Msys2\tdm32\bin\libgmodule-2.0-0.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000068100000 \\?\I:\Msys2\tdm32\bin\libgobject-2.0-0.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006B040000 \\?\I:\Msys2\tdm32\bin\libffi-8.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000076820000 \\?\C:\Windows\SysWOW64\shlwapi.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000071470000 \\?\C:\Windows\SysWOW64\IPHLPAPI.DLL
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006ECE0000 \\?\C:\Windows\SysWOW64\dnsapi.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000063080000 \\?\I:\Msys2\tdm32\bin\zlib1.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000077170000 \\?\C:\Windows\SysWOW64\imm32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000077300000 \\?\C:\Windows\SysWOW64\nsi.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006ED70000 \\?\C:\Windows\SysWOW64\winhttp.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000726C0000 \\?\C:\Windows\SysWOW64\cryptsp.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000072690000 \\?\C:\Windows\SysWOW64\rsaenh.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000077310000 \\?\C:\Windows\SysWOW64\bcrypt.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000072680000 \\?\C:\Windows\SysWOW64\cryptbase.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075500000 \\?\C:\Windows\SysWOW64\bcryptprimitives.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000074650000 \\?\C:\Windows\SysWOW64\kernel.appcore.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000073BD0000 \\?\C:\Windows\SysWOW64\uxtheme.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000760D0000 \\?\C:\Windows\SysWOW64\clbcatq.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006D770000 \\?\C:\Windows\SysWOW64\AppXDeploymentClient.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006AE50000 \\?\C:\Windows\SysWOW64\Windows.ApplicationModel.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000075840000 \\?\C:\Windows\SysWOW64\SHCore.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000771A0000 \\?\C:\Windows\SysWOW64\oleaut32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000031C0000 \\?\C:\Windows\SysWOW64\twinapi.appcore.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000070080000 \\?\C:\Windows\SysWOW64\windows.storage.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000070050000 \\?\C:\Windows\SysWOW64\wldp.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=00000000753C0000 \\?\C:\Windows\SysWOW64\cfgmgr32.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006F420000 \\?\C:\Windows\SysWOW64\propsys.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000071B10000 \\?\C:\Windows\SysWOW64\profapi.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006ACA0000 \\?\C:\Windows\SysWOW64\mssprxy.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006ABE0000 \\?\C:\Windows\SysWOW64\MrmCoreR.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006EF40000 \\?\C:\Windows\SysWOW64\iertutil.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006DAD0000 \\?\C:\Windows\SysWOW64\Windows.StateRepositoryCore.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006AAE0000 \\?\C:\Windows\SysWOW64\Windows.UI.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006E4D0000 \\?\C:\Windows\SysWOW64\TextInputFramework.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006E1B0000 \\?\C:\Windows\SysWOW64\CoreUIComponents.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006A970000 \\?\C:\Windows\SysWOW64\WindowManagementAPI.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006A9F0000 \\?\C:\Windows\SysWOW64\InputHost.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006E430000 \\?\C:\Windows\SysWOW64\CoreMessaging.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=0000000071510000 \\?\C:\Windows\SysWOW64\ntmarta.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006E0D0000 \\?\C:\Windows\SysWOW64\WinTypes.dll
LOAD_DLL PID=25104 TID=27140 lpBaseOfDll=000000006A940000 \\?\C:\Windows\SysWOW64\BCP47mrm.dll
CREATE_THREAD PID=25104 TID=18760
EXCEPTION PID=25104 TID=18760 ExceptionCode=0x80000003 dwFirstChance=1
EXIT_THREAD PID=25104 TID=18760 dwExitCode=0x0
EXCEPTION PID=25104 TID=27140 ExceptionCode=0xc0000005 dwFirstChance=0
gdk-pixbuf-query-loaders.exe caused an Access Violation at location 00000000672A87D0 Reading from location 00000000672A87D0.

Registers:
eax=bf09abd8 ebx=00831ce8 ecx=672a87d0 edx=fc26af62 esi=672a87d0 edi=00831cf4
eip=672a87d0 esp=0060fe88 ebp=0060fec4 iopl=0         ov up ei pl nz na pe cy
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00210a03

AddrPC   Params
672A87D0 FEF9D206 0083CA04 00000030
75936608 00000000 00000000 00000000  msvcrt.dll!_initterm_e+0x158
759366A1 00000000 0083C9F0 00831568  msvcrt.dll!exit+0x11
00401493 00000000 7542FA29 002AA000  gdk-pixbuf-query-loaders.exe!0x1493
7542FA29 002AA000 FC28334A 00000000  KERNEL32.DLL!BaseThreadInitThunk+0x19
774A7C3E FFFFFFFF 774C8DD1 00000000  ntdll.dll!RtlGetAppContainerNamedObjectPath+0x11e
774A7C0E 004014B0 002AA000 00000000  ntdll.dll!RtlGetAppContainerNamedObjectPath+0xee

EXIT_THREAD PID=25104 TID=30424 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=23088 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=26772 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=3744 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=17764 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=20260 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=22420 dwExitCode=0xc0000005
EXIT_THREAD PID=25104 TID=19412 dwExitCode=0xc0000005
EXIT_PROCESS PID=25104 TID=27140 dwExitCode=0xc0000005

```

output from drmingw on running gdk-pixbuf-query-loaders with my version (sometimes still fails as you can see strangely no pointer in this version to gmutex_lock/unlock).

gdb output:

```
(gdb) run
Starting program: I:\Msys2\tdm32\bin\gdk-pixbuf-query-loaders.exe
[New Thread 16828.0x64a4]
[New Thread 16828.0x55a4]
[New Thread 16828.0x5954]
[New Thread 16828.0x3384]

Thread 1 received signal SIGSEGV, Segmentation fault.
0x6244a566 in ?? () from I:\Msys2\tdm32\bin\libyuv.dll
(gdb) continue
Continuing.

Thread 1 received signal SIGSEGV, Segmentation fault.
0x6244a566 in ?? () from I:\Msys2\tdm32\bin\libyuv.dll

```

removing the libyuv.dll then progresses further but fails on libpixbufloader-svg.dll and a few others.

**Reply by revelator** [reply]
lldb output ->
```

Process 23384 launched: 'I:\Msys2\tdm32\bin\gdk-pixbuf-query-loaders.exe' (i386)
Process 23384 stopped
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x6244a566: Access violation reading location 0x707730a4
    frame #0: 0x6244a566 libyuv.dll`CanonicalFourCC + 52166
libyuv.dll`CanonicalFourCC:
->  0x6244a566 <+52166>: movl   0x4(%eax), %eax
    0x6244a569 <+52169>: testl  %eax, %eax
    0x6244a56b <+52171>: jne    0x6244a558                ; <+52152>
    0x6244a56d <+52173>: cmpl   $0x624580a0, %edx         ; imm = 0x624580A0
```

**Reply by revelator** [reply]
btw i was not trying to stop the mutex from doing what it was supposed to, i was trying to avoid anything else from accessing it when it was doing it. Ok got the original version bootstrapped but i also have to rebuild the dependencies of gdk-pixbuf2 which will probably take a while since the Msys2 project has grown into a monster of circular dependencies xD. All my work was actually to get it accepted as a plugin for the Msys2 shell since most gcc devs on windows use that atm (well except the hardcore users who prefer doing it from wsl).

**Reply by revelator** [reply]
Oo omg. i just discovered the source of the segfault and you are going to laugh now (or maybe cry i was not sure either) it is caused by 
this -> if ( WaitForSingleObject( yada, yada, yada)  == waitforit) drumrool.... if i change the "if" to "while" it works even with the old version what the actuall hell....

**Reply by revelator** [reply]
seems a bug to me but using "while" actually works i created a small test program using the code from your version only with the atomic counter i used in mine as an extra guard (does not hurt anything).

```
mutex lock1 signaled: check error walue (null)
mutex lock1 running: count was: (00000001): check error walue (null)
mutex lock2 signaled: check error walue (null)
mutex lock2 running: count was: (00000001): check error walue (null)
```

this is the output, i use a small Sleep timer to simulate running code after we get the lock.

so it does what it is supposed to now it seems, the error null just means there are no errors returned from GetLastError.
attaching the source and test executable it is actually pretty helpfull to muck around with it, try for instance to change the "while" to "if" again and build it with debugging on.

I remove the acl code, though it works with that also, so feel free to use it (was actully from a microsoft dev in regards to getting full access to a mutex object on win10).

**Reply by revelator** [reply]
[test-mutex-src.zip](https://github.com/jmeubank/tdm-gcc/files/9742550/test-mutex-src.zip)

here is the test program and source, does work in bash also but you need to press ctrl+C to see the output if run directly it shows instantly.

**Comment by revelator** [comment]
Debug build of gcc-10.4.0 with the TDM patchset uploaded here -> https://sourceforge.net/projects/cbadvanced/files/TDM%20Debug/

built with -ggdb3 to get as much debug info as possible, changed the if before WaitForSIngleObject to while, all static dependencies used.
built with binutils-2.39 but since the 32 bit build is sjlj it should not matter, if it had been a dwarf build it would be required though due to the new dwarf 5 format, the 64 bit build is SEH.

**Reply by revelator** [reply]
turns out the mutex lock was not working at all in either function, lots of debugging later and i now have a working gcc-11.3.0-tdm :) with full ASLR and DEP support, could not get ASLR working previously in TDM turned out it was the mutex lock that keeled us over.

so far i managed to build 50% of the MSYS2 mingw packages with it (damn thing has grown to a monster xD) so it will be a while but i can upload the packages when done. Ill also include replacement packages for MSYS2's filesystem and pacman to add the TDM compiler to the toolchain. 

@jmeubank feel free to use my patch and PKGBUILD when it is up. 
btw. in case you were wondering im RalphE so not the first time i done some work for the stability of this toolchain. i retired from the coding scene some years back though so im a bit rusty.

**Reply by revelator** [reply]
![TDM](https://user-images.githubusercontent.com/1930128/198894763-d0a324eb-c01b-4678-af33-98ce625f3f22.jpg)

graphviz built by TDM with QT5

**Reply by revelator** [reply]
https://sourceforge.net/projects/cbadvanced/files/TDM%20Release%20ASLR/

compiler and binutils no external dependencies.

still building a frag ton of MSYS2 mingw-w64 packages with this which will be uploaded later. there is a chance that this package might make it into the MSYS2 release if approved.
